src/report.rs

```diff
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Intel SGX Enclave report structures.
+
+use crate::parameters::{Attributes, Features, MiscSelect, Xfrm};
+
+/// The enclave report body.
+///
+/// For more information see the following documents:
+///
+/// [Intel® Software Guard Extensions (Intel® SGX) Data Center Attestation Primitives: ECDSA Quote Library API](https://download.01.org/intel-sgx/dcap-1.0/docs/SGX_ECDSA_QuoteGenReference_DCAP_API_Linux_1.0.pdf)
+///
+/// Table 5, A.4. Quote Format
+///
+/// [Intel® 64 and IA-32 Architectures Software Developer's Manual Volume 3 (3A, 3B, 3C & 3D): System Programming Guide](https://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-software-developer-vol-3d-part-4-manual.html)
+///
+/// Table 38-21. Layout of REPORT
+#[derive(Clone, Debug)]
+#[repr(C)]
+pub struct ReportBody {
+    cpusvn: [u8; 16],
+    miscselect: [u8; 4],
+    reserved1: [u8; 28],
+    // The attributes field is composed of the features and xfrm fields.
+    features: [u8; 8],
+    xfrm: [u8; 8],
+    mrenclave: [u8; 32],
+    reserved2: [u8; 32],
+    mrsigner: [u8; 32],
+    reserved3: [u8; 96],
+    isv_prodid: [u8; 2],
+    isv_svn: [u8; 2],
+    reserved4: [u8; 60],
+    reportdata: [u8; 64],
+}
+
+impl ReportBody {
+    /// The SVN (security version number) of the processor.
+    pub fn cpu_security_version(&self) -> [u8; 16] {
+        self.cpusvn
+    }
+
+    /// Bit vector specifying which extended features are saved to the MISC region of the
+    /// SSA frame when an AEX occurs.
+    ///
+    /// If it cannot be parsed the raw little endian bytes will be returned instead.
+    pub fn misc_select(&self) -> Result<MiscSelect, [u8; 4]> {
+        let misc_select = u32::from_le_bytes(self.miscselect);
+        MiscSelect::from_bits(misc_select).ok_or(self.miscselect)
+    }
+
+    /// Set of flags describing attributes of the enclave.
+    ///
+    /// If it cannot be parsed the raw little endian bytes will be returned instead.
+    ///
+    /// The raw bytes returned are the 64 bit features and xfrm respectively.
+    pub fn attributes(&self) -> Result<Attributes, ([u8; 8], [u8; 8])> {
+        let features = u64::from_le_bytes(self.features);
+        let features = Features::from_bits(features).ok_or((self.features, self.xfrm))?;
+        let xfrm = u64::from_le_bytes(self.xfrm);
+        let xfrm = Xfrm::from_bits(xfrm).ok_or((self.features, self.xfrm))?;
+        Ok(Attributes::new(features, xfrm))
+    }
```

Comment:
@definitelynobody I've been thinking about this more. Passing the raw bytes as an error might be useful for debugging. But it could also disclose information about the attestation if not careful.

I think we should use `T::from_bits_truncate()` and avoid an error here altogether.